src/rust/core/server/src/threads/worker/multi.rs

```diff
@@ -0,0 +1,351 @@
+                        }
+                    }
+                }
+
+                if session.read_pending() > 0
+                    && Self::handle_session_read(session, &self.poll, &mut self.storage_queue)
+                {
+                    self.wake_storage = true;
+                }
+            }
+        }
+    }
+
+    fn handle_new_sessions(&mut self) {
+        while let Ok(mut session) = self.session_queue.recv_from(0) {
+            let pending = session.read_pending();
+            trace!("{} bytes pending in rx buffer for new session", pending);
+
+            // reserve vacant slab
+            let session_entry = self.sessions.vacant_entry();
+            let token = Token(session_entry.key());
+
+            // set client token to match slab
+            session.set_token(token);
+
+            // register tcp stream and insert into slab if successful
+            match session.register(&self.poll) {
+                Ok(_) => {
+                    session_entry.insert(session);
+                    if pending > 0 {
+                        // handle any pending data immediately
+                        if self.handle_data(token).is_err() {
+                            self.handle_error(token);
+                        }
+                    }
+                }
+                Err(_) => {
+                    error!("Error registering new socket");
+                }
+            };
+        }
+    }
+
+    pub fn session_sender(&mut self, waker: Arc<Waker>) -> QueuePair<Session, ()> {
+        self.session_queue.new_pair(65536, Some(waker))
+    }
+
+    pub fn signal_queue(&mut self) -> QueuePair<Signal, ()> {
+        self.signal_queue.new_pair(128, None)
+    }
+}
+
+impl<Storage, Request, Response> EventLoop for MultiWorker<Storage, Request, Response>
+where
+    Request: Parse + Send,
+    Response: Compose + Send,
+    Storage: Execute<Request, Response> + EntryStore + Send,
+{
+    fn get_mut_session(&mut self, token: Token) -> Option<&mut Session> {
+        self.sessions.get_mut(token.0)
+    }
+
+    fn handle_data(&mut self, token: Token) -> Result<(), ()> {
+        trace!("handling request for session: {}", token.0);
+        if let Some(session) = self.sessions.get_mut(token.0) {
+            if session.write_capacity() > 0
+                && Self::handle_session_read(session, &self.poll, &mut self.storage_queue)
+            {
+                self.wake_storage = true;
+            }
+
+            if session.write_pending() > 0 && session.flush().is_ok() && session.write_pending() > 0
+            {
+                self.reregister(token);
+            }
+
+            Ok(())
+        } else {
+            // no session for the token
+            trace!(
+                "attempted to handle data for non-existent session: {}",
+                token.0
+            );
+            Ok(())
+        }
+    }
+
+    /// Reregister the session given its token
+    fn reregister(&mut self, token: Token) {
```

Comment:
Are `reregister` &  `take_session` the same across all threads? Should it be put into the `EventLoop` pub trait ?

Reply:
Yeah - might be able to share an impl if we add another simple required function for the trait too